Add Custom FPS Limiter

makotech222 opened the pull request:
Addresses issue https://github.com/Ryujinx/Ryujinx/issues/4164

![image](https://user-images.githubusercontent.com/4389156/236718993-ca2f9123-d6cb-40a9-8f1a-44d93c5597c5.png)

![image](https://user-images.githubusercontent.com/4389156/236719037-c43ea5a6-deef-42f0-93e0-75c9931db2b1.png)

![image](https://user-images.githubusercontent.com/4389156/236720087-03e0e475-6e9e-48aa-bea7-7b07a9b6d168.png)



Taking design cues from duckstation and pcsx2, I've added 3 states for the fps limit:
 * Normal
 * Fast Forward
 * Turbo

This allows user to define a custom emulation speed while playing the game. The names "Normal, Fast Forward, Turbo" are not special in any technical way; the user can configure them to be any speed. The naming convention is just inherited from other emulators to provide consistent user experience.

The speeds are stored as a Decimal, and are multiplied against the expected normal emulation speed of 60 fps (even if a game runs at a lower fps).

Two new hotkeys are added to toggle Fast Forward and Turbo. If nothing is toggled, then speed will be in normal state. If both are toggled on, then turbo will override fast forward.

Vsync functionality has changed. It is no longer used to limit emulation speed. A speed limit of Unlimited (-1) will replicate the functionality that was previously covered by disabling vsync, for purpose of emulation speed and simulating Switch functionality. Vsync is still used by the Window classes and the Vulkan renderer, but I'm not entirely sure those do anything functionally. Vsync label has been removed from the bottom status bar and replaced with the current SpeedState.

Future Work not in this PR:
* Add code path for avalonia gui


On a side note, absolute major kudos to ryujinx dev team; I was able to clone repo and build project in first try, and find the place where I needed to make changes within 2 minutes. Absolutely marvelously structured project, great job!

al81-ru commented:
Hi! What about games that run at 30 or 20 (ha!) fps?

al81-ru commented:
Hi! I would suggest leave vsynced mode as it is, as it does frameskipping when under target framerate. And instead limit unlocked mode

makotech222 commented:
> Hi! I would suggest leave vsynced mode as it is, as it does frameskipping when under target framerate. And instead limit unlocked mode

I'm doubtful it does frameskipping, but if you set your speed to 60 fps, it will do the same thing that enabling vsync does.

al81-ru commented:
> I'm doubtful it does frameskipping, but if you set your speed to 60 fps, it will do the same thing that enabling vsync does.

Corrected, not frame skipping, but frame delaying. Compose() may not post frame if it is not ready, but will wait another _ticksPerFrame to try again

emmauss commented:
Vsync value is used in the village window to enable immediate present mode, which is necessary on supported GPUs to bypass/disable driver vsync. It's the equivalent of swap interval for opengl.

emmauss commented:
Also, those 3 states don't truly work as you would think in ryujinx. There is no fast forward or high speed mode. It's completely up to the game as to how and when to send frames, even when we tell the game that a frame is always ready. Some games adjust game play speed to the current frame rate, effectively running at 100% game speed, even when frame rate drops. Those options don't so what they are meant to do. Other games run their game loop around the swap rate. In those games, we can have fast forward and turbo modes.

makotech222 commented:
Any example games of that behavior? In the games i tested, it worked perfectly.

al81-ru commented:
> Any example games of that behavior?

Luigi's Mansion 3 relies on TickSource's ticks, it maintains 30 fps with internal waitings

emmauss commented:
Celeste for instance, always run at full game speed, even when frame rate drops to like 20 and looks like a slide show. Game physics run at their own pace

makotech222 commented:
So disabling vsync doesn't have any effect on the games? In that case, there's likely not much we can do outside of someone writing a mod it seems like? In any case, it seems like most games respond perfectly to it, so no reason to not implement it.

emmauss commented:
> So disabling vsync doesn't have any effect on the games? In that case, there's likely not much we can do outside of someone writing a mod it seems like? In any case, it seems like most games respond perfectly to it, so no reason to not implement it.

Both games behave differently with vsync off. Celeste will simply generate as many frames as possible, but game physics run at a steady rate.
How different is fast forward and turbo, from each other and vsync off currently?

gdkchan commented:
I think it makes more sense to add a slider where you can set any target frame rate that you want. There was a request for that on #4164 (I think?).

makotech222 commented:
Yeah the GUI portion isn't done, but it will be user customizable limits. Currently, just set the numbers in the config file.

MutantAura commented:
Worth mentioning that the numbers themselves won't be fully accurate to what the user may think the setting does. As emmauss sort of explained games can do different things and while a 60FPS game will work as expected, a game targeting 30FPS won't see those values. Instead they'll see exactly half of the target.

Currently this PR acts as more of a "framerate target multiplier" which is fine, but isn't really what the requester had in mind.

makotech222 commented:
@MutantAura That is correct, the terminology can probably be reworded to something like what duckstation does, which is a percentage of full speed (150%, 100%, etc)

emmauss commented:
> @MutantAura That is correct, the terminology can probably be reworded to something like what duckstation does, which is a percentage of full speed (150%, 100%, etc)

What work 150% speed be for a 30fps game?

emmauss commented:
If possible, could you drop in on our dev channel on discord, for much easier discussion.

riperiperi commented:
Just a few things:

- This is adjusting the refresh rate of the device, rather than a frame cap.
- Vsync off (and swapInterval 0 behaviour, which _is_ present on switch) should not be removed. They allow the framerate to be completely uncapped, with the vsync event coinciding with the frame presentation. This is useful for fully dynamic framerates in certain games, and benchmarking in others.

MutantAura commented on src/Ryujinx.HLE/Switch.cs:
```diff
@@ -127,6 +142,49 @@ public void EnableCheats()
             FileSystem.ModLoader.EnableCheats(Processes.ActiveApplication.ProgramId, TamperMachine);
         }
 
+        public void SetSpeedState(SpeedState newState)
```
Left a general codestyle message on Discord.

makotech222 replied:
Okay checked in some changes to address codestyle. take another look and let me know if you want me to do any more

MutantAura commented on src/Ryujinx.Ui.Common/Configuration/ConfigurationState.cs:
```diff
@@ -749,6 +773,9 @@ public void LoadDefault()
             Graphics.ShadersDumpPath.Value            = "";
+            Graphics.NormalEmulationSpeed.Value       = 1.0m;
+            Graphics.FastForwardEmulationSpeed.Value           = 1.5m;
+            Graphics.TurboEmulationSpeed.Value                 = -1.0m;
```
Misaligned.

MutantAura commented on src/Ryujinx.Ui.Common/Configuration/ConfigurationState.cs:
```diff
@@ -1409,6 +1464,9 @@ public ConfigurationLoadResult Load(ConfigurationFileFormat configurationFileFor
             Graphics.ScalingFilterLevel.Value         = configurationFileFormat.ScalingFilterLevel;
+            Graphics.NormalEmulationSpeed.Value       = configurationFileFormat.NormalEmulationSpeed;
+            Graphics.FastForwardEmulationSpeed.Value           = configurationFileFormat.FastForwardEmulationSpeed;
+            Graphics.TurboEmulationSpeed.Value                 = configurationFileFormat.TurboEmulationSpeed;
```
Misaligned

MutantAura commented on src/Ryujinx/Ui/MainWindow.glade:
```diff
@@ -698,15 +698,15 @@
                             <property name="margin-left">5</property>
                             <property name="margin-right">5</property>
-                            <property name="label" translatable="yes">VSync</property>
+                            <property name="label" translatable="yes">Normal</property>
```
Not sure I like the "Normal", "Fast-Forward", "Turbo" terminology too much although I've never used Duckstation or PCSX2. Personally I would simply leave it as "VSync", "VSync (XYZ%)", "Unlocked" which I think better describe to our current userbase what these do. Would also allow us to maintain the colouring, which I also quite like!

More thoughts on this from others would be good though.

makotech222 replied:
With your terminology, it implies the first and third options are not configurable, which wouldn't be as desirable. I'd welcome a better terminology, though, if anyone has any suggestions.

Sentox6 replied:
Speaking from a user perspective, I'm not sure they should be configurable. Personally, I'd like to see a Normal, Custom, and Unlocked setting, with only the second being, well, customisable. I do think this would better communicate the actual behaviour of the setting than the current VSync nomenclature.

If per-game settings are eventually implemented, this would become significantly more useful, of course.

makotech222 replied:
I disagree. As a user who loves this feature in other emulators, I customize all three and swap between them pretty commonly as I play. It's pretty easy to understand what it does.

Sentox6 replied:
If that's the desired functionality, then the user should be able to add and label any number of custom speeds as they see fit, rather than three arbitrarily-named settings.

MutantAura reviewed (commented):
Functionality wise this works as expected. Left a few comments.

Would be preferable to also implement Avalonia side before merge as it is the priority for all new features.

makotech222 commented:
Added change to include Custom values in the GTK UI:

![image](https://github.com/Ryujinx/Ryujinx/assets/4389156/5c6715d5-e545-467f-b18e-144a64d80816)

I think I've implemented all that I wanted to. @MutantAura is Avalonia support going to be a requirement before merge? I'd like to take that on at a later date, since its not the official gui yet.

MutantAura commented:
> I think I've implemented all that I wanted to. @MutantAura is Avalonia support going to be a requirement before merge? I'd like to take that on at a later date, since its not the official gui yet.

On this front I'm not sure. Ideally we don't want them to diverge on features and if they do then Avalonia should be prioritized. 
Most of the legwork is already done here, if you don't have the time to finish that side then I can possibly add it in a follow up or push to the branch...

TSRBerry commented:
This needs a rebase, a workflow should have timed out here, but didn't.

makotech222 commented:
Okay rebased it

makotech222 commented:
Added avalonia GUI:

![image](https://github.com/Ryujinx/Ryujinx/assets/4389156/9f72533f-3b58-4ded-9edf-199161d34d80)

![image](https://github.com/Ryujinx/Ryujinx/assets/4389156/abd19e51-b3bd-4ef1-9181-64488b5b6092)

jcm93 commented:
Just tried to test this out on macOS but it seems not to be working, is that just because it's not wired up with Avalonia (Avalonia being the only available GUI on Mac)? Looks cool, looking forward to testing it.

jcm93 commented:
> I tried to test it on Ava (macOS), but it seems to have no effect yet.
> If you type in a custom value, there is an x shown while the field is active.
> Because of that you can't actually see, what you are typing.

I had a comment earlier that I deleted until I could do a proper followup.

I was testing on macOS and also thought that it wasn't working. It turns out that it is, but the changes only take effect after restarting the title. You should see the changes reflected after restarting the game.

I agree that the input is a mess, not sure if it's specific to macOS. The custom input asking for a decimal number is also odd, given that the dropdown has percent values. Should be uniform across both I would think.

siegmund-heiss-ich commented:
True.. I didn't think about it..
Nothing in `System` can be changed while a game is running.
Normal emulation speed works fine.
But i can't toggle the other two options.
I assigned the Hotkeys correctly but it didn't work for me.
I also tested it with resolution scaling to be sure I did nothing wrong.
Also if I disable vsync, the game is locked at `Normal Emulation Speed`.

makotech222 commented:
Sorry guys, didn't test avalonia as much as I should have. All those issues should now be fixed! Please try testing again.

siegmund-heiss-ich commented:
Sadly all the described problems are still there on macOS.
I can't toggle any other speed while emulating a game.

Also the custom field should also has values in %:
<img width="716" alt="Bildschirmfoto 2023-05-30 um 12 51 36" src="https://github.com/Ryujinx/Ryujinx/assets/119589995/e13773aa-70c0-41fc-bfdd-86cf1f95a755">

This `x` is quite annoying. It appears, when the input field is active:
<img width="701" alt="Bildschirmfoto 2023-05-30 um 12 51 46" src="https://github.com/Ryujinx/Ryujinx/assets/119589995/59c78f6f-5308-40b8-aaa8-9031221cc78e">

But great work! Keep it up!

siegmund-heiss-ich commented:
I deleted my last comment because I didn't test the latest version.
Sorry for that.

Now everything seems to work great on Avalonia.

My last request would be, to show values in % if you want to set a custom value.
At the Moment it looks like this:
<img width="758" alt="Bildschirmfoto 2023-05-30 um 13 12 31" src="https://github.com/Ryujinx/Ryujinx/assets/119589995/4f6eef46-fd81-4dff-94ae-d3b85d56c80d">

makotech222 commented:
Okay, i've changed the custom value to be more like a percent to match dropdown options (multiplied decimal by 100)

MutantAura commented:
Functionality wise it's all working good on my end. I still have reservations about a few mainly UI related things though.

1. The naming system I still really dislike. They simply do not align with how Switch games work. Fast-Forward and Turbo make no sense in literally thousands of games. They won't fast forward, nor speed up at all. Plenty of first party titles also have dynamic framerates and so this entire aspect of naming feels stuck in the past and non-applicable to modern emulation. I would still really suggest naming things _what they do_. What is happening here is simply changing the FPS target for present interval, so simply "VSync (X%)" is way more accurate, correlates with what users already know, and is easy to see in screenshots for troubleshooting or sharing purposes.

2. This breaks the VSync toggle hotkey. F1/TAB depending on when the config file was created. Should probably be fixed to either cycle through the options or simply flit between what currently is called "normal" and "turbo". 

3. Keep the green/red colours for the widget text. At this point I think they're part of the program identity and to lose them would make me sad :( 

I would appreciate if anyone else could chime in on my first point as it's obviously subjective.

MutantAura commented on src/Ryujinx.Ava/UI/ViewModels/SettingsViewModel.cs:
```diff
@@ -409,6 +483,16 @@ public void LoadCurrentConfiguration()
             ExpandDramSize = config.System.ExpandRam;
             IgnoreMissingServices = config.System.IgnoreMissingServices;
 
+            var foundNormalSpeed = _emulationSpeeds.Values.Contains(config.Graphics.NormalEmulationSpeed) ? _emulationSpeeds.First(x => x.Value == config.Graphics.NormalEmulationSpeed).Key : null;
+            NormalEmulationSpeed = foundNormalSpeed != null ? AvailableNormalSpeeds.IndexOf(AvailableNormalSpeeds.First(x => (string)x.Content == foundNormalSpeed)) : AvailableNormalSpeeds.IndexOf(AvailableNormalSpeeds.First(x => (string)x.Content == _emulationSpeeds.First(x => x.Value == -2).Key));
+            CustomNormalEmulationSpeed = config.Graphics.NormalEmulationSpeed * 100.0m;
+            var foundFastForwardSpeed = _emulationSpeeds.Values.Contains(config.Graphics.FastForwardEmulationSpeed) ? _emulationSpeeds.First(x => x.Value == config.Graphics.FastForwardEmulationSpeed).Key : null;
+            FastForwardSpeed = foundFastForwardSpeed != null ? AvailableFastForwardSpeeds.IndexOf(AvailableFastForwardSpeeds.First(x => (string)x.Content == foundFastForwardSpeed)) : AvailableFastForwardSpeeds.IndexOf(AvailableFastForwardSpeeds.First(x => (string)x.Content == _emulationSpeeds.First(x => x.Value == -2).Key));
+            CustomFastForwardSpeed = config.Graphics.FastForwardEmulationSpeed * 100.0m;
+            var foundTurboSpeed = _emulationSpeeds.Values.Contains(config.Graphics.TurboEmulationSpeed) ? _emulationSpeeds.First(x => x.Value == config.Graphics.TurboEmulationSpeed).Key : null;
+            TurboSpeed = foundTurboSpeed != null ? AvailableTurboSpeeds.IndexOf(AvailableTurboSpeeds.First(x => (string)x.Content == foundTurboSpeed)) : AvailableTurboSpeeds.IndexOf(AvailableTurboSpeeds.First(x => (string)x.Content == _emulationSpeeds.First(x => x.Value == -2).Key));
+            CustomTurboSpeed = config.Graphics.TurboEmulationSpeed * 100.0m;
```
Split these out for readability, may also wish to define constants for 100.0m and -2.

MutantAura commented on src/Ryujinx.Ava/UI/ViewModels/MainWindowViewModel.cs:
```diff
@@ -393,17 +393,6 @@ public Brush ProgressBarForegroundColor
             }
         }
 
-        public Brush VsyncColor
-        {
-            get => _vsyncColor;
-            set
-            {
-                _vsyncColor = value;
-
-                OnPropertyChanged();
-            }
-        }
-
```
I like the colours :(

MutantAura reviewed (commented):
Couple of design issues to resolve, mainly the abundant use of some magic decimals which are not instantly obvious. GUI layout and design should be shifted to glade and perhaps think about your hotkey design because at the moment there is some toggle vsync code still lying around not really doing much.

jcm93 commented:
I was trying to rebase this for a custom build and it seems like the Avalonia combo boxes that bind to ``AvailableNormalSpeeds``, ``AvailableFastForwardSpeeds`` are broken, presumably by the recent Avalonia update, just a word of warning to anyone trying to make builds off of this PR.

makotech222 commented:
Okay i've rebased again. Can we please get one more approval on this? I don't have much resources to continue maintaining this PR as things change over time. We can always iterate on the feature over time after it is merged as well.

TSRBerry commented:
I'm not sure what you are asking here.

1. The format workflow is still failing.
2. You didn't even address most of the review comments.

> We can always iterate on the feature over time after it is merged as well.

This can't be merged in the current state. Open review comments generally mean there at least needs to be some discussion and maybe also a few changes. We don't want to merge PRs if there are still concerns about them, especially if there weren't discussed yet.

> I don't have much resources to continue maintaining this PR as things change over time.

I understand the time issue of course and get that it's not easy to keep your PR up to date all the time. Especially, since we are really slow at reviewing things. That's why I like to rebase for people if I notice the PR isn't up to date and we have taken quite some time to review it already.
I'd do the same with your PR if all checks were succeeding and review comments addressed.

makotech222 commented:
Sorry i was being terse, just tired at the moment.

Format workflow failing is news to me. I'll take a look at that when i have time.
Forgot those review comments were there from mutant also. Will try to look when i have time

jcm93 commented:
The following are just some usability opinions from having used this PR a fair amount on a Switch title that features tech that involves specific FPS values.

Integrating with the VSync option should probably be done. In its current state it's fairly confusing to hit a button to change the target refresh rate to 200%, but still be at 60 Hz because VSync is still enabled. I agree that it would make the most sense to use the existing VSync toggle hotkey to switch between the options in order, instead of having specific hotkeys for "fast forward" and "turbo". Especially because the behavior is somewhat confusing (i.e. if you're on fast forward, and hit the turbo hotkey twice, are you back at fast forward? or at normal?) At a minimum, incorporating the VSync option and hotkey seems good.

I agree with MutantAura's comments about the naming, but I don't see a super easy solution. Perhaps the "fast forward" and "turbo" names could be kept in the Settings dialog, but in the Ryujinx status bar use the suggested "VSync %" and show the value? I'd personally prefer nameable presets by the user, but that's just my opinion. I also think it's clearer to use number refresh rate values instead of percentages, but again, just my opinion. Overall the PR seems to work quite well and I certainly find it very useful and I thank you a lot for implementing it!

jcm93 commented:
Hey there,

I wanted to check and see if this had been abandoned or not. If you don't have enough time to see it across the finish line, I was going to offer to take it over. I don't want to step on your toes if you are going to finish though. Let me know!

jcm93 commented:
OK, sounds good. Thanks for all the work!

StealthTranslations commented:
> Hey there,
> 
> I wanted to check and see if this had been abandoned or not. If you don't have enough time to see it across the finish line, I was going to offer to take it over. I don't want to step on your toes if you are going to finish though. Let me know!

Did you take over the project?

jcm93 commented:
Yes, I should have tagged it, the new PR is at https://github.com/Ryujinx/Ryujinx/pull/5735.
